website/static/js/fangorn.js

```diff
@@ -88,31 +89,70 @@ function findByTempID(parent, tmpID) {
-    var rejected = tb.dropzone.getRejectedFiles();
-    var queued = tb.dropzone.getQueuedFiles();
-    var filesArr = uploading.concat(queued, rejected);
-
-    for (var i = 0; i < filesArr.length; i++) {
-        var j = filesArr[i];
-        if(!row) {
-            var parent = j.treebeardParent || tb.dropzoneItemCache;
-            var item = findByTempID(parent, j.tmpID);
-            tb.dropzone.removeFile(j);
-            tb.deleteNode(parent.id,item.id);
-        } else {
-            tb.deleteNode(row.parentID,row.id);
-            if(row.data.tmpID === j.tmpID){
-                tb.dropzone.removeFile(j);
+    var cancelableStatuses = [Dropzone.UPLOADING, Dropzone.QUEUED];
+    // Select files that are uploading, queued, or rejected (!accepted)
+    var filesArr = tb.dropzone.files.filter(function(file) {
+        return cancelableStatuses.indexOf(file.status) > -1 || !file.accepted;
+    });
+    // Remove all queued files
+    if (row === undefined) {
+        var removeFromUI = function(file) {
+            var parent = file.treebeardParent || tb.dropzoneItemCache;
+            var item = findByTempID(parent, file.tmpID);
+            tb.deleteNode(parent.id, item.id);
+        };
+        // Clear all synchronous uploads
+        SYNC_UPLOAD_ADDONS.forEach(function(provider) {
+            if (tb.dropzone.syncFileCache[provider] !== undefined) {
+                // Remove cached provider files from UI
+                tb.dropzone.syncFileCache[provider].forEach(removeFromUI);
+                // Clear provider cache
+                tb.dropzone.syncFileCache[provider].length = 0;
             }
-        }
+        });
+        // Clear all ongoing uploads
+        filesArr.forEach(function(file, index) {
+            // Ignore completed files
+            if (file.upload.progress === 100) return;
+            removeFromUI(file);
+            // Cancel currently uploading file
+            tb.dropzone.removeFile(file);
+        });
+        tb.isUploading(false);
+        return;
+    }
+    var handled = false;
+    // Search for and remove specified file from queue
+    if (SYNC_UPLOAD_ADDONS.indexOf(row.data.provider) !== -1) {
+        // Provider is handled in sync
+        handled = tb.dropzone.syncFileCache[row.data.provider].some(function(file, index) {
+            if (file.tmpID === row.data.tmpID) {
+                tb.deleteNode(row.parentID, row.id);
+                return tb.dropzone.syncFileCache[row.data.provider].splice(index, 1);
+            }
+        });
+    }
+    if (!handled) {
+        // File is currently being uploaded/managed by dropzone
+        handled = filesArr.some(function(file) {
+            if (file.tmpID === row.data.tmpID) {
+                tb.deleteNode(row.parentID, row.id);
+                tb.dropzone.removeFile(file);
+                return true;
+            }
+        });
     }
-    tb.isUploading(false);
+    tb.isUploading(handled && filesArr.length > 1);
 }
```

Comment:
This method is getting unwieldy. Let's go ahead and split it into two methods, one for canceling all uploads, and one for canceling one upload.